// @pushprotocol/restapi/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PACKAGE_BUILD = exports.ALPHA_FEATURES = exports.MessageType = exports.ENCRYPTION_TYPE = exports.ENV = void 0;
// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageJson = require('../../package.json');
/**
 * SUPPORTED ENVIRONEMENTS
 */
var ENV;
(function (ENV) {
    ENV["PROD"] = "prod";
    ENV["STAGING"] = "staging";
    ENV["DEV"] = "dev";
    /**
     * **This is for local development only**
     */
    ENV["LOCAL"] = "local";
})(ENV = exports.ENV || (exports.ENV = {}));
/**
 * SUPPORTED ENCRYPTIONS FOR PUSH PROFILE
 */
var ENCRYPTION_TYPE;
(function (ENCRYPTION_TYPE) {
    ENCRYPTION_TYPE["PGP_V1"] = "x25519-xsalsa20-poly1305";
    ENCRYPTION_TYPE["PGP_V2"] = "aes256GcmHkdfSha256";
    ENCRYPTION_TYPE["PGP_V3"] = "eip191-aes256-gcm-hkdf-sha256";
    ENCRYPTION_TYPE["NFTPGP_V1"] = "pgpv1:nft";
})(ENCRYPTION_TYPE = exports.ENCRYPTION_TYPE || (exports.ENCRYPTION_TYPE = {}));
/**
 * SUPPORTED MESSAGE TYPES FOR PUSH CHAT
 */
var MessageType;
(function (MessageType) {
    MessageType["TEXT"] = "Text";
    MessageType["IMAGE"] = "Image";
    MessageType["VIDEO"] = "Video";
    MessageType["AUDIO"] = "Audio";
    MessageType["FILE"] = "File";
    /** @deprecated - Use `MediaEmbed` Instead */
    MessageType["GIF"] = "GIF";
    MessageType["MEDIA_EMBED"] = "MediaEmbed";
    MessageType["META"] = "Meta";
    MessageType["REACTION"] = "Reaction";
    MessageType["RECEIPT"] = "Receipt";
    MessageType["USER_ACTIVITY"] = "UserActivity";
    MessageType["INTENT"] = "Intent";
    MessageType["REPLY"] = "Reply";
    MessageType["COMPOSITE"] = "Composite";
    MessageType["PAYMENT"] = "Payment";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
exports.ALPHA_FEATURES = {
    SCALABILITY_V2: 'SCALABILITY_V2',
};
exports.PACKAGE_BUILD = packageJson.version.includes('alpha')
    ? 'ALPHA'
    : 'STABLE';
const Constants = {
    ENV,
    ENCRYPTION_TYPE,
    PAGINATION: {
        INITIAL_PAGE: 1,
        LIMIT: 10,
        LIMIT_MIN: 1,
        LIMIT_MAX: 50,
    },
    DEFAULT_CHAIN_ID: 11155111,
    DEV_CHAIN_ID: 99999,
    NON_ETH_CHAINS: [137, 80001, 56, 97, 10, 420, 1442, 1101, 421613, 42161, 122, 123],
    ETH_CHAINS: [1, 11155111],
    ENC_TYPE_V1: 'x25519-xsalsa20-poly1305',
    ENC_TYPE_V2: 'aes256GcmHkdfSha256',
    ENC_TYPE_V3: 'eip191-aes256-gcm-hkdf-sha256',
    ENC_TYPE_V4: 'pgpv1:nft',
    ALPHA_FEATURES: exports.ALPHA_FEATURES,
};
exports.default = Constants;
