// @pushprotocol/restapi/src/lib/pushapi/profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Profile = void 0;
const tslib_1 = require("tslib");
const PUSH_USER = require("../user");
const PushAPI_1 = require("./PushAPI");
class Profile {
    constructor(account, env, decryptedPgpPvtKey, progressHook) {
        this.account = account;
        this.env = env;
        this.decryptedPgpPvtKey = decryptedPgpPvtKey;
        this.progressHook = progressHook;
    }
    info(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const accountToUse = (options === null || options === void 0 ? void 0 : options.overrideAccount) || this.account;
            const response = yield PUSH_USER.get({
                account: accountToUse,
                env: this.env,
            });
            return response.profile;
        });
    }
    update(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.decryptedPgpPvtKey) {
                throw new Error(PushAPI_1.PushAPI.ensureSignerMessage());
            }
            const { name, desc, picture } = options;
            const response = yield PUSH_USER.profile.update({
                pgpPrivateKey: this.decryptedPgpPvtKey,
                account: this.account,
                profile: {
                    name: name,
                    desc: desc,
                    picture: picture,
                },
                env: this.env,
                progressHook: this.progressHook,
            });
            return response.profile;
        });
    }
}
exports.Profile = Profile;
