// @pushprotocol/socket/src/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVENTS = exports.DEFAULT_CHAINS = exports.ENV = void 0;
var ENV;
(function (ENV) {
    ENV["PROD"] = "prod";
    ENV["STAGING"] = "staging";
    ENV["DEV"] = "dev";
    /**
     * **This is for local development only**
     */
    ENV["LOCAL"] = "local";
})(ENV = exports.ENV || (exports.ENV = {}));
var DEFAULT_CHAINS;
(function (DEFAULT_CHAINS) {
    DEFAULT_CHAINS[DEFAULT_CHAINS["PROD"] = 1] = "PROD";
    DEFAULT_CHAINS[DEFAULT_CHAINS["STAGING"] = 11155111] = "STAGING";
    DEFAULT_CHAINS[DEFAULT_CHAINS["DEV"] = 11155111] = "DEV";
    /**
     * **This is for local development only**
     */
    DEFAULT_CHAINS[DEFAULT_CHAINS["LOCAL"] = 11155111] = "LOCAL";
})(DEFAULT_CHAINS = exports.DEFAULT_CHAINS || (exports.DEFAULT_CHAINS = {}));
exports.EVENTS = {
    // Websocket
    CONNECT: 'connect',
    DISCONNECT: 'disconnect',
    // Notification
    USER_FEEDS: 'userFeeds',
    USER_SPAM_FEEDS: 'userSpamFeeds',
    // Chat
    CHAT_RECEIVED_MESSAGE: 'CHATS',
    CHAT_GROUPS: 'CHAT_GROUPS'
};
