// @livepeer/core/dist/livepeer-core.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var errors_dist_livepeerCoreErrors = require('../errors/dist/livepeer-core-errors.cjs.dev.js');
var client = require('./client-7e4fd284.cjs.dev.js');
var version = require('./version-dcf1673a.cjs.dev.js');
var updateStream = require('./updateStream-26465265.cjs.dev.js');
var getStream = require('./getStream-a1e88460.cjs.dev.js');
var provider = require('./provider-56b1ec78.cjs.dev.js');
var controller = require('./controller-3e1ecbbf.cjs.dev.js');
var string = require('./string-0759efb1.cjs.dev.js');
var deepMerge = require('./deepMerge-cb091110.cjs.dev.js');
var omick = require('./omick-042aa9a4.cjs.dev.js');
require('zustand/middleware');
require('zustand/vanilla');



exports.ACCESS_CONTROL_ERROR_MESSAGE = errors_dist_livepeerCoreErrors.ACCESS_CONTROL_ERROR_MESSAGE;
exports.BFRAMES_ERROR_MESSAGE = errors_dist_livepeerCoreErrors.BFRAMES_ERROR_MESSAGE;
exports.HttpError = errors_dist_livepeerCoreErrors.HttpError;
exports.NOT_ACCEPTABLE_ERROR_MESSAGE = errors_dist_livepeerCoreErrors.NOT_ACCEPTABLE_ERROR_MESSAGE;
exports.STREAM_OFFLINE_ERROR_MESSAGE = errors_dist_livepeerCoreErrors.STREAM_OFFLINE_ERROR_MESSAGE;
exports.STREAM_OPEN_ERROR_MESSAGE = errors_dist_livepeerCoreErrors.STREAM_OPEN_ERROR_MESSAGE;
exports.isAccessControlError = errors_dist_livepeerCoreErrors.isAccessControlError;
exports.isBframesError = errors_dist_livepeerCoreErrors.isBframesError;
exports.isNotAcceptableError = errors_dist_livepeerCoreErrors.isNotAcceptableError;
exports.isStreamOfflineError = errors_dist_livepeerCoreErrors.isStreamOfflineError;
exports.Client = client.Client;
exports.clearClient = client.clearClient;
exports.createClient = client.createClient;
exports.createStorage = client.createStorage;
exports.noopStorage = client.noopStorage;
exports.version = version.version;
exports.createAsset = updateStream.createAsset;
exports.createClip = updateStream.createClip;
exports.createStream = updateStream.createStream;
exports.getAsset = updateStream.getAsset;
exports.getAssetMetrics = updateStream.getAssetMetrics;
exports.getPlaybackInfo = updateStream.getPlaybackInfo;
exports.getStreamSession = updateStream.getStreamSession;
exports.getStreamSessions = updateStream.getStreamSessions;
exports.updateAsset = updateStream.updateAsset;
exports.updateStream = updateStream.updateStream;
exports.watchLivepeerProvider = updateStream.watchLivepeerProvider;
exports.getLivepeerProvider = getStream.getLivepeerProvider;
exports.getStream = getStream.getStream;
exports.defaultStudioConfig = provider.defaultStudioConfig;
exports.defaultTranscodingProfiles = provider.defaultTranscodingProfiles;
exports.createControllerStore = controller.createControllerStore;
exports.getMediaSourceType = controller.getMediaSourceType;
exports.sanitizeMediaControllerState = controller.sanitizeMediaControllerState;
exports.b64Decode = string.b64Decode;
exports.b64Encode = string.b64Encode;
exports.b64UrlDecode = string.b64UrlDecode;
exports.b64UrlEncode = string.b64UrlEncode;
exports.deepMerge = deepMerge.deepMerge;
exports.omit = omick.omit;
exports.pick = omick.pick;
