// @livepeer/core/dist/updateStream-7eba57a6.cjs.prod.js
'use strict';

var getStream = require('./getStream-03a26cdc.cjs.prod.js');
var client = require('./client-b49d615d.cjs.prod.js');
var provider = require('./provider-68cbe4e6.cjs.prod.js');

function watchLivepeerProvider(callback) {
  const client$1 = client.getClient();
  const handleChange = async () => callback(getStream.getLivepeerProvider());
  const unsubscribe = client$1.subscribe(({
    provider
  }) => provider, handleChange);
  return unsubscribe;
}

function createAsset(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.createAsset(args);
}

function getAsset(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.getAsset(args);
}

function updateAsset(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.updateAsset(args);
}

function getAssetMetrics(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.getAssetMetrics(args);
}

function getPlaybackInfo(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.getPlaybackInfo(args);
}

function createClip(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.createClip(args);
}

function createStream(args) {
  const provider$1 = getStream.getLivepeerProvider();
  return provider$1.createStream({
    ...args,
    profiles: args?.profiles ?? provider.defaultTranscodingProfiles
  });
}

function getStreamSession(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.getStreamSession(args);
}

function getStreamSessions(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.getStreamSessions(args);
}

function updateStream(args) {
  const provider = getStream.getLivepeerProvider();
  return provider.updateStream(args);
}

exports.createAsset = createAsset;
exports.createClip = createClip;
exports.createStream = createStream;
exports.getAsset = getAsset;
exports.getAssetMetrics = getAssetMetrics;
exports.getPlaybackInfo = getPlaybackInfo;
exports.getStreamSession = getStreamSession;
exports.getStreamSessions = getStreamSessions;
exports.updateAsset = updateAsset;
exports.updateStream = updateStream;
exports.watchLivepeerProvider = watchLivepeerProvider;
