// @pushprotocol/restapi/src/lib/space/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = void 0;
const tslib_1 = require("tslib");
const constants_1 = require("../constants");
const helpers_1 = require("./../chat/helpers");
const createGroup_1 = require("../chat/createGroup");
function create(options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { signer, spaceName, spaceDescription, listeners, spaceImage, speakers, isPublic, contractAddressNFT, numberOfNFTs, contractAddressERC20, numberOfERC20, env = constants_1.default.ENV.PROD, pgpPrivateKey = null, scheduleAt, scheduleEnd, rules, } = options || {};
        const groupRules = rules ? (0, helpers_1.convertSpaceRulesToRules)(rules) : null;
        try {
            const group = yield (0, createGroup_1.createGroup)({
                signer,
                groupName: spaceName,
                groupDescription: spaceDescription,
                members: listeners,
                groupImage: spaceImage,
                admins: speakers,
                isPublic: isPublic,
                contractAddressNFT: contractAddressNFT,
                numberOfNFTs: numberOfNFTs,
                contractAddressERC20: contractAddressERC20,
                numberOfERC20: numberOfERC20,
                env,
                pgpPrivateKey,
                groupType: 'spaces',
                scheduleAt: scheduleAt,
                scheduleEnd: scheduleEnd,
                rules: groupRules,
            });
            return (0, helpers_1.groupDtoToSpaceDto)(group);
        }
        catch (err) {
            console.error(`[Push SDK] - API  - Error - API ${create.name} -:  `, err);
            throw Error(`[Push SDK] - API  - Error - API ${create.name} -: ${err}`);
        }
    });
}
exports.create = create;
