// @pushprotocol/restapi/src/lib/space/removeSpeakers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeSpeakers = void 0;
const tslib_1 = require("tslib");
const constants_1 = require("../constants");
const helpers_1 = require("../chat/helpers");
const removeAdmins_1 = require("../chat/removeAdmins");
const removeSpeakers = (options) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const { spaceId, speakers, signer = null, env = constants_1.default.ENV.PROD, pgpPrivateKey = null, } = options || {};
    try {
        const group = yield (0, removeAdmins_1.removeAdmins)({
            chatId: spaceId,
            admins: speakers,
            signer: signer,
            env: env,
            pgpPrivateKey: pgpPrivateKey,
        });
        return (0, helpers_1.groupDtoToSpaceDtoV2)(group, env);
    }
    catch (err) {
        console.error(`[Push SDK] - API  - Error - API ${exports.removeSpeakers.name} -:  `, err);
        throw Error(`[Push SDK] - API  - Error - API ${exports.removeSpeakers.name} -: ${err}`);
    }
});
exports.removeSpeakers = removeSpeakers;
